### PR TITLE
SAA-484: Use appointment categories from NOMIS

### PR DESCRIPTION
**Pull request by iterative-it**

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/ReferenceCodeService.kt [hunk 1]
```diff
@@ -0,0 +1,28 @@
+package uk.gov.justice.digital.hmpps.hmppsactivitiesmanagementapi.service
+
+import org.springframework.stereotype.Service
+import uk.gov.justice.digital.hmpps.hmppsactivitiesmanagementapi.client.prisonapi.api.PrisonApiClient
+import uk.gov.justice.digital.hmpps.hmppsactivitiesmanagementapi.client.prisonapi.model.ReferenceCode
+
+@Service
+class ReferenceCodeService(
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/ReferenceCodeService.kt:8]: Not suggesting you change it, this just lit a light-bulb (for me anyway). I do wonder if we should have one dedicated service for all reference type data.  

Just conscious the number of services is growing and some do very little.

e.g. activities categories, attendance reasons etc could go in here?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/ReferenceCodeService.kt [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+  private val prisonApiClient: PrisonApiClient,
+) {
+  fun getReferenceCodes(domain: String): List<ReferenceCode> =
+    prisonApiClient.getReferenceCodes(domain)
+
+  fun getAppointmentCategoryReferenceCodes(): List<ReferenceCode> =
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/ReferenceCodeService.kt:14]: You could reduce the number of methods here.  One that takes in an enum perhaps to determine the ref code type you are after?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/AppointmentService.kt [hunk 3]
```diff
@@ -109,14 +105,16 @@ class AppointmentService(
...
+      .getOrDefault(categoryCode, null)
+      ?: throw IllegalArgumentException("Appointment Category with code $categoryCode not found or is not active")
   }
 
   private fun failIfLocationNotFound(request: AppointmentCreateRequest) {
     if (!request.inCell) {
-      locationService.getLocationsForAppointments(request.prisonCode!!)
-        ?.firstOrNull { location -> location.locationId == request.internalLocationId }
+      locationService.getLocationsForAppointmentsMap(request.prisonCode!!)
+        .getOrDefault(request.internalLocationId, null)
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/AppointmentService.kt:117]: You shouldn't need the getOrDefault on the maps like this, this should have the same effect.
```
    referenceCodeService.getAppointmentScheduleReasonsMap()[categoryCode]
      ?: throw IllegalArgumentException("Appointment Category with code $categoryCode not found or is not active")

```

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/util/TransformFunctions.kt [hunk 4]
```diff
@@ -456,12 +462,21 @@ fun List<AppointmentInstance>.toScheduledEvent(
...
-    eventSubType = it.category.code,
-    eventSubTypeDesc = it.category.description,
+    eventSubType = category.code,
+    eventSubTypeDesc = category.description,
     agencyId = it.prisonCode,
   )
 }
 
+fun ReferenceCode?.toAppointmentCategorySummary(code: String) =
+  if (this == null) {
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/util/TransformFunctions.kt:472]: Just a suggestion (I presume both code and description would be null at the same time):

```
fun ReferenceCode?.toAppointmentCategorySummary(code: String) =
  AppointmentCategorySummary(this?.code ?: code, this?.description ?: "UNKNOWN")

```

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/ReferenceCodeServiceTest.kt [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+        listOf(
+          appointmentCategoryReferenceCode("AC1", "Appointment Category 1"),
+          appointmentCategoryReferenceCode("AC2", "Appointment Category 2"),
+          appointmentCategoryReferenceCode("AC3", "Appointment Category 3"),
+        ),
+      )
+
+    val locations = referenceCodeService.getReferenceCodes("INT_SCH_RSN")
+
+    Assertions.assertThat(locations).containsExactly(
```
Comment (matt-crowson-moj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/ReferenceCodeServiceTest.kt:28]: Whoops the old `Assertions.` classic.

**Review by matt-crowson-moj — APPROVED** [review]
LGTM only very minor comments.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/ReferenceCodeService.kt (hunk 2 above)
Comment (iterative-it) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/ReferenceCodeService.kt]: I really like this suggestion and have implemented it

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/util/TransformFunctions.kt (hunk 4 above)
Comment (iterative-it) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/util/TransformFunctions.kt:472]: The passed in code is the code the entity record has and won't be null. The idea here is the UI could still display the code that was originally used along with the "UNKNOWN" description

**Review by matt-crowson-moj — APPROVED** [review]
LGTM

**Review by matt-crowson-moj — APPROVED** [review]
LGTM

**Pull request merged**